src/System.Net.WebHeaderCollection/src/System/Net/WebHeaderCollection.cs

```diff
@@ -134,85 +134,90 @@ private bool AllowHttpResponseHeader
+                        }
+                        else if (c == 127 || (c < ' ' && c != '\t'))
+                        {
+                            throw new ArgumentException(SR.Format(SR.net_WebHeaderInvalidControlChars, nameof(value)), nameof(value));
+                        }
+                        break;
+
+                    case 1:
+                        if (c == '\n')
+                        {
+                            crlf = 2;
                             break;
+                        }
+                        throw new ArgumentException(SR.Format(SR.net_WebHeaderInvalidCRLFChars, nameof(value)), nameof(value));
 
-                        case 1:
-                            if (c == '\n')
-                            {
-                                crlf = 2;
-                                break;
-                            }
-                            throw new ArgumentException(SR.Format(SR.net_WebHeaderInvalidCRLFChars, "value"));
-
-                        case 2:
-                            if (c == ' ' || c == '\t')
-                            {
-                                crlf = 0;
-                                break;
-                            }
-                            throw new ArgumentException(SR.Format(SR.net_WebHeaderInvalidCRLFChars, "value"));
-                    }
-                }
-                if (crlf != 0)
-                {
-                    throw new ArgumentException(SR.Format(SR.net_WebHeaderInvalidCRLFChars, "value"));
+                    case 2:
+                        if (c == ' ' || c == '\t')
+                        {
+                            crlf = 0;
+                            break;
+                        }
+                        throw new ArgumentException(SR.Format(SR.net_WebHeaderInvalidCRLFChars, nameof(value)), nameof(value));
                 }
             }
-            else
+
+            if (crlf != 0)
             {
-                // NAME check
-                // First, check for absence of separators and spaces
-                if (HttpValidationHelpers.IsInvalidMethodOrHeaderString(name))
-                {
-                    throw new ArgumentException(SR.Format(SR.net_WebHeaderInvalidHeaderChars, "name"));
-                }
+                throw new ArgumentException(SR.Format(SR.net_WebHeaderInvalidCRLFChars, nameof(value)), nameof(value));
+            }
 
-                // Second, check for non CTL ASCII-7 characters (32-126)
-                if (ContainsNonAsciiChars(name))
-                {
-                    throw new ArgumentException(SR.Format(SR.net_WebHeaderInvalidNonAsciiChars, "name"));
-                }
+            return value;
+        }
+
+        /// <summary>
+        /// Throws on invalid header name chars.
+        /// </summary>
+        private static string CheckBadHeaderNameChars(string name)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(name));
```

Comment:
Why did you replace the throw with an assert?  Can't a null legitimately come in via ```this[whatever] = null;```? https://github.com/dotnet/corefx/pull/7289/files?diff=split#diff-485d01bd1bd24491e498fce07a89f759R112